### ISSUE TITLE
Make a utility taking systems into AMBER/GROMACS

### ISSUE THREAD

**Issue by davidlmobley** [open-forcefield-group/smarty#184]
Make a utility or example which will take SMIRFF system and go to AMBER `prmtop` and `crd` and/or GROMACS `top` and `gro`, as per https://github.com/choderalab/yank/issues/544#issuecomment-262032469

I can rely on github.com/mobleylab/solvationtoolkit for system prep, so really this means just taking a system which is already prepped (`.pdb` and `.mol2` files for the system and components available), parameterizing with SMIRFF, and going into GROMACS or AMBER format. Since the conversion from OpenMM to AMBER/GROMACS can be handled by `ParmEd`, I think this should be fairly trivial.

**Comment by jchodera** [comment]
Do we want to use TIP3P water or parm@frosst water? And would we want to use Amber or SMIRFF formats for the water?

You'll want to include as input:
* PDB file for the system to read in Topology and positions
* Mol2 files with partial charges for small molecule and solcent (if parameterizing solvent box)

I would save the serialized OpenMM System XML as well, just in case that is useful.

**Comment by davidlmobley** [comment]
@jchodera - how would I save the serialized OpenMM System XML? Am not familiar with that.

Without some new development, I think we can only do AlkEthOH flexible water if we take the whole system from SMIRFF. We don't have constraints yet, nor do we have TIP3P in SMIRFF.

Potentially I could figure out how to do a SMIRFF solute in conventional TIP3P if I bring in TIP3P from AMBER (which would also allow for constraints), though this will take me longer. As a first pass implementation I plan to just provide a way to take an arbitrary all-SMIRFF system into AMBER and GROMACS.

**Comment by jchodera** [comment]
Actually, I wonder if we can easily graft in a reader for OpenMM XML and PDB (much like the amber or Gromacs readers)  so you can skip the ParmEd step. Let me ask @andrrizzi in the morning.

**Comment by jchodera** [comment]
> @jchodera - how would I save the serialized OpenMM System XML? Am not familiar with that.

```python
# serialize system to a string
from simtk.openmm import XmlSerializer
serialized_system = XmlSerializer.serialize(system)
# now write that string to a file
```

> Without some new development, I think we can only do AlkEthOH flexible water if we take the whole system from SMIRFF. We don't have constraints yet, nor do we have TIP3P in SMIRFF.

We can probably easily add constraints. Did that make it into the spec, or
is there just an issue?

We could also encode TIP3P water in SMIRFF if we wanted to use that instead.

> Potentially I could figure out how to do a SMIRFF solute in conventional TIP3P if I bring in TIP3P from AMBER (which would also allow for constraints), though this will take me longer. As a first pass implementation I plan to just provide a way to take an arbitrary all-SMIRFF system into AMBER and GROMACS.

Sounds good. I'd focus on the first part (prior to ParmEd) in case we can
add that XML reader input.

On the other hand, it may be useful to the paper how you can use SMIRFF in
Gromacs or Amber, so that conversion could be valuable.

**Comment by davidlmobley** [comment]
> We could also encode TIP3P water in SMIRFF if we wanted to use that instead.

Yes, that's a good point. The major issue is just the constraints.

I'll go ahead with the converter generally as this is something we need for other reasons, aside from just Yank.

**Comment by andrrizzi** [comment]
> I wonder if we can easily graft in a reader for OpenMM XML and PDB (much like the amber or Gromacs readers) so you can skip the ParmEd step

Sorry for the late response! Yes, that's something that could be implemented in a day or two.

**Comment by davidlmobley** [comment]
Willing to do so soon?  That would help a ton and allow things to be much less amber centric.

**Comment by andrrizzi** [comment]
Ok! I'll do my best to get it in by the end of the week.

**Comment by andrrizzi** [comment]
Not sure if you received the notification via email @davidlmobley, but I've added the feature to support XML systems.

**Comment by davidlmobley** [comment]
I did, @andrrizzi ! I have it on my short list to review. However, I've not used Yank at all yet, so it may take me a bit to figure out how to do this. What would help a lot would be if you could point me something that would show how to take a prepared system (i.e. via this support) and do a hydration calculation. Do you have an example? Even an implicit solvent hydration calculation would be a good start, though I'd also want to do explicit. THat would certainly expedite my review. :) 

Thanks.

**Comment by andrrizzi** [comment]
Of course! No pressure, I just wanted to be sure that you knew.

@Lnaden has written some [great tutorials](http://getyank.org/latest/examples/index.html) that explain in details the examples in the [yank-example repo](https://github.com/choderalab/yank-examples/). That includes hydration free energy examples from SMILES and a `mol2` file, and it's a great place to get the general idea. Unfortunately, we don't have examples documentation yet for specifically setting up a calculation from system files (rather than molecules files that go through the automatic pipeline), but this is how the script would look like:
```yaml
---
options:
  minimize: yes
  verbose: yes
  number_of_iterations: 1000
  temperature: 300*kelvin
  pressure: 1*atmosphere

systems:
  smarty-system:
    phase1_path: [solventsystem.pdb, solventsystem.xml]
    phase2_path: [vacuumsystem.pdb, vacuumsystem.xml]
    ligand_dsl: resname MOL

protocols:
  hydration-protocol:
    solvent1:
      alchemical_path:
        lambda_electrostatics: [1.00, 0.75, 0.50, 0.25, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00]
        lambda_sterics:        [1.00, 1.00, 1.00, 1.00, 1.00, 0.95, 0.90, 0.85, 0.80, 0.75, 0.70, 0.65, 0.60, 0.50, 0.40, 0.30, 0.20, 0.10, 0.00]
    solvent2:
      alchemical_path:
        lambda_electrostatics: [1.00, 0.75, 0.50, 0.25, 0.00]
        lambda_sterics:        [1.00, 1.00, 1.00, 1.00, 1.00]

experiments:
  system: smarty-system
  protocol: hydration-protocol
```

**Comment by davidlmobley** [comment]
@andrrizzi - how about if I want to do a hydration calculation for just a single molecule in GBSA solvent if I bring in the molecule parameters from an XML file? i.e. like the phenol implicit solvent example in yank-examples, but where I load phenol parameters from XML? Can this be done, or does it require something beyond what's currently supported?

Thanks.

**Comment by jchodera** [comment]
@davidlmobley : Some questions:
* Are you only targeting GBSA here? And why? Is this just debugging, or do you plan to report these in the SMIRFF paper? If you just want to do GBSA, we can write a 10-line script to do that instead.
* The XML file will need to contain the assigned GBSA parameters. I'll need to finish this PR:
https://github.com/open-forcefield-group/smarty/pull/182

**Comment by davidlmobley** [comment]
Just debugging, @jchodera -- wanted something I could run really quickly to verify things are behaving before plunging on to systems consisting of many molecules which will run much more slowly.

**Comment by andrrizzi** [comment]
Hmm. The parameters should already be included in the system `xml` file that you feed to the script, so you shouldn't need to specify them explicitly. Or am I missing something?

**Comment by davidlmobley** [comment]
@andrrizzi - I think there are two issues. The first is what I was originally asking -- I just don't know your yaml syntax enough yet to know how to adapt the example you sent to turn it into an implicit solvent case. Your example loads two entirely prepared systems; the example in `yank-examples` loads a molecule and solvates in (water, vacuum). What I want to do is something in between (if possible) -- load a molecule + parameters from an (XML, pdb) and then solvate in (water, vacuum). 

But, I think @jchodera is right in that probably, the issue is that I would need to add the GBSA solvent before dumping the system out to XML in the first place, which would mean we'd need GBSA parameters in our `smirff` format.

@jchodera - what are your plans on finishing #182?

**Comment by jchodera** [comment]
I'd give the explicit solvent calculations a try, @davidlmobley. You should be able to add the TIP3P water parameters in SMIRFF format, at least for flexible TIP3P. If we're good with https://github.com/open-forcefield-group/smarty/issues/100 we can implement that too.

**Comment by jchodera** [comment]
> @andrrizzi - I think there are two issues. The first is what I was originally asking -- I just don't know your yaml syntax enough yet to know how to adapt the example you sent to turn it into an implicit solvent case. Your example loads two entirely prepared systems; the example in yank-examples loads a molecule and solvates in (water, vacuum). What I want to do is something in between (if possible) -- load a molecule + parameters from an (XML, pdb) and then solvate in (water, vacuum).

You would need to provide XML and PDB files for both "molecule with `GBSAOBCForce`" and "molecule in vacuum". YANK won't add any forces or GBSA models or solvents if you are using this input route.

**Comment by davidlmobley** [comment]
Yes, we're good with #100, @jchodera .

**Comment by jchodera** [comment]
@davidlmobley : What's most critical, GBSA support (for implicit calculations) or constraint support (for explicit solvent TIP3P)?

**Comment by andrrizzi** [comment]
> What I want to do is something in between (if possible) -- load a molecule + parameters from an (XML, pdb) and then solvate in (water, vacuum).

Yeah, sorry, this would require creating a pipeline that is based on XML files, and it would take some time to implement. The only current way to do this is to provide two different `xml` files with their own parameters.

**Comment by davidlmobley** [comment]
Constraint support is a higher priority for me, @jchodera, though I was under the impression you had to have GBSA support in right away for your other work.

**Comment by davidlmobley** [comment]
@andrrizzi :
> Yeah, sorry, this would require creating a pipeline that is based on XML files, and it would take some time to implement. The only current way to do this is to provide two different xml files with their own parameters.

OK, no problem. I just thought this would be a super simple way to test, but I will instead try and proceed towards systems of many molecules (i.e. phenol in some solvent).

**Comment by davidlmobley** [comment]
@andrrizzi - OK, so I have my systems prepared (I think) and am trying to use this. However, I'm running into an error relating to my gas phase system -- it says that a receptor-ligand system requires a restraint?? (I'm trying to run the solvation free energy for a solute in cyclohexane and in gas).

Here's the error:
```
Traceback (most recent call last):
  File "/anaconda/bin/yank", line 11, in <module>
    load_entry_point('yank==0.14.1', 'console_scripts', 'yank')()
  File "//anaconda/lib/python3.5/site-packages/yank/cli.py", line 74, in main
    dispatched = getattr(commands, command).dispatch(command_args)
  File "//anaconda/lib/python3.5/site-packages/yank/commands/script.py", line 59, in dispatch
    yaml_builder.build_experiments()
  File "//anaconda/lib/python3.5/site-packages/yank/yamlbuild.py", line 1298, in build_experiments
    self._run_experiment(combination, output_dir)
  File "//anaconda/lib/python3.5/site-packages/yank/yamlbuild.py", line 2462, in _run_experiment
    yank.create(thermodynamic_state, *phases, restraint_type=restraint_type)
  File "//anaconda/lib/python3.5/site-packages/yank/yank.py", line 286, in create
    self._create_phase(thermodynamic_state, phase, restraint_type)
  File "//anaconda/lib/python3.5/site-packages/yank/yank.py", line 454, in _create_phase
    raise ValueError('A receptor-ligand system in implicit solvent or '
ValueError: A receptor-ligand system in implicit solvent or vacuum requires a restraint.
```

Files are attached.

[SAMPL5_003.yaml.zip](https://github.com/open-forcefield-group/smarty/files/643259/SAMPL5_003.yaml.zip)
[inputfiles.zip](https://github.com/open-forcefield-group/smarty/files/643260/inputfiles.zip)

**Comment by andrrizzi** [comment]
@davidlmobley Thanks for catching this!

Yank guesses which atoms are part of the ligand from the input ligand DSL, which ones are part of the solvent based on the common solvent residue names used by mdtraj, and put as receptor everything else. Since the octane molecules in your input files don't use any of [these residue names](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/core/residue_names.py#L23-L42), Yank classify them as part of the receptor.

We need to catch this error earlier, and make stuff more flexible. For now, a simple workaround is to use one of those residue names for your solvent molecules.

**Comment by andrrizzi** [comment]
Also, the `NonbondedForce` in `cyclohexane-SAMPL5_003.xml` uses `NoCutoff` as nonbonded method, so Yank thinks that you want to simulate in implicit solvent or in vacuum. I think you need to fix that in the xml preparation code.

**Comment by davidlmobley** [comment]
Yes, @andrrizzi - having an enumerated list of acceptable residue names that are recognized as "solvent" is not an acceptable solution to this problem, as one might need to simulate hundreds or thousands of solvents. Also, AMBER files require that each molecule have a unique residue name -- so if I want to parameterize, say, 100 different small molecules and simulate them in various combinations, there is going to be no easy way to make this fit with a "every solvent must have one of these residue names" model (your list is not long enough).

**Comment by davidlmobley** [comment]
Resolved by #187 .
